쥬스 메이커 [STEP 3] safari, Red

Siwon-L opened the pull request:
@stevenkim18
안녕하세요 steven 😁 😃
2_group의 safari, Red 입니다. 잘 부탁드립니다!
쥬스메이커 STEP 3 PR 보냅니다
![](https://i.imgur.com/LQpivZf.jpg)



## 문제상황과 해결방법
- 모달을 활용한 뷰 이동에서 중간 네비게이션컨트롤러 때문에, 만든 뷰 인스턴스와 실제 이동된 뷰의 인스턴스가 달랐던 문제
- 화면간 데이터 이동에 관한 문제

### 모달을 활용한 뷰 이동에서 중간 네비게이션컨트롤러 때문에, 만든 뷰 인스턴스와 실제 이동된 뷰의 인스턴스가 달랐던 문제
```swift
    private func setUpDelegate(_ ManagingStockView: ManagingStockViewController) {
        ManagingStockView.stock = juiceMaker.fruitStore.stock
        ManagingStockView.delegate = self
    }
    
    private func moveManagingStockView() {
        guard let ManagingStockViewNavigation = self.storyboard?.instantiateViewController(withIdentifier: "ManagingStockViewNavigation") as? UINavigationController else { return }
        guard let ManagingStockView = ManagingStockViewNavigation.topViewController as? ManagingStockViewController else { return }
        setUpDelegate(ManagingStockView)
        self.present(ManagingStockViewNavigation, animated: true, completion: nil)
    }
```
처음 `MainViewController의` 인스턴스를 직접 만들어 재고에 대한 데이터를 입력한 뒤, 화면을 이동시켰는데 데이터가 입력되있지 않았습니다. 확인을 해보니 내비게이션 컨트롤을 거쳐 이동되는 과정에서 실제 보여지는 뷰의 인스턴스와 데이터를 넣어주기 위해 만든 인스턴스가 독립된 인스턴스였고, 내비게이션 컨트롤러의 탑뷰 컨트롤러가 사용자에게 보이는 화면인 것을 알 수 있었습니다.


### 화면간 데이터 이동에 관한 문제
step2에서 네비게이션 컨트롤러를 이용한 데이터 이동을 시도해 보았으나, 현제 프로젝트에서는 여러 오브젝트 간의 일대다 관계가 필요하지 않다고 판단하여 네비게이션 센터를 제거하였습니다. 
`ManagingStockViewController` 에서 `MainViewController`로 값을 가져올 경우에는 delegate를 사용하여 일을 위임하면서, 함수에 파라미터로 값을 넘겨주는 방식을 선택했습니다. 이 방식이 데이터도 넘겨주면서 일은 `MainViewController` 에서 시킬 수 있기 때문에, 가져와야 할 정보가 줄어들어 용의하다고 판단했습니다.

## 고민한 점 & 조언을 받고 싶은 부분
- 컨트롤러 부분이 너무 방대해 지는 것에 대해서
- Stepper의 사이즈가 줄어들지 않는 것에 대해서

### 컨트롤러 부분이 너무 방대해 지는 것에 대해서 
처음에 코드에 대한 계획을 세울 때에는 모델에서 최대한 많은 일을 처리하고, 컨트롤러에서는 최소한의 관리를 해주는 것으로 생각했습니다. 그러나 코드를 짜다보니, 컨트롤러에서 뷰의 라이프 사이클을 관리하고 있어서 컨트롤러의 코드의 양을 줄이기가 쉽지 않았습니다. 원래 컨트롤러가 이렇게 커져도 되는 것일까요? 혹은, 관리하는 데이터의 양이 적어 자연스러운 현상인지 아니면 더 작게 관리할 수 있는 방법이 있는 걸까요? 

### Stepper의 사이즈가 줄어들지 않는 것에 대해서
현재 프로젝트는 portrait 모드를 막아 놓아서 세로로 화면을 보지 못합니다. 
그런데 `MainViewController` 의 화면을 세로로 볼 경우에는 stepper 들의 크기가 겹치는 현상이 나타나고 있습니다. stepper 사이즈의 기본값이 설정 되어 있어 조절 할 수가 없었습니다. 혹시 세로로 화면을 보아야 한다면, 어떻게 처리해야 할까요? stepper 사이즈를 줄이거나, 아니면 stack view 자체를 두줄로 내려오게 만드는 등의 방법이 있는지 궁금합니다.

cherrishRed committed:
Merge branch 'STEP2'

Commit:
refactor : initLable함수 네이밍 변경

initLable -> initializeLable

cherrishRed committed:
refactor : updateFruitLable에서 노티피케이션 삭제

Commit:
refactor : notification post 삭제 및 initializeLable()함수 삭제

Fruit타입 CaseIterable 프로토콜 탈락

Commit:
refactor : notification옵져버 삭제 및 initializeLable()함수 삭제

Fruit타입 CaseIterable 프로토콜 탈락

cherrishRed committed:
Merge branch 'STEP3' of https://github.com/saafaaari/ios-juice-maker into STEP3

cherrishRed committed:
refactor : viewWillAppear 제거

orderJuice 에서 updateFruitLable 호출

Commit:
add :  SubViewController파일 추가

재고 추가 View UINavigationItem 추가 및 barButton 추가

cherrishRed committed:
refactor : closeView 함수 구현

Commit:
refactor : SubViewController 옵셔널 바인딩

cherrishRed committed:
feat : SubViewController 5개 라벨 생성

Commit:
feat : moveStockChangeView함수 뷰간 데이터 전달 기능 추가

cherrishRed committed:
feat : SubViewController 가 보일 때 라벨 값 업데이트

Commit:
feat : Stepper를 이용한 딸기 재고 수정 기능 추가

updateLabel()함수 Stepper.value초기값 설정 기능 추가

cherrishRed committed:
feat : Stepper를 이용한 모든 재고 수정 기능 추가

Commit:
feat : delegate 패턴을 위한 Update 프로토콜 추가

SubViewController타입의 delegate변수 추가

cherrishRed committed:
feat : 버튼을 누를 때마다, stockNumbers 업데이트 구현

Commit:
feat : 재고 수정 뷰가 종료될때 데이터 전달 기능 추가

stockChangeView.delegate 설정(ViewController)
updateModel함수 기능 구현
FruitStore타입 내부 updateStock메소드 추가

cherrishRed committed:
feat : 오토레이아웃을 위한 객채 묶기

Commit:
feat : 오토레이아웃 제약 추가

cherrishRed committed:
refactor : Update 프로토콜 네이밍 Updateable 로 변경

Updateable 안의 메서드 네이밍 변경
updateFruitLable을 updateStockLable으로 네이밍 변경

Commit:
refactor : 네이밍, 파라미터 전달인자 수정 및 함수 분리

stockNumbers -> stock
update(stock: [Fruit: Int]) -> update(for stock: [Fruit: Int])
updateLabel() -> initializeViewValue() 및 viewDidLoad() 이동
initializeViewValue함수에서 스위치문 initializeViewValue함수로 분리

cherrishRed committed:
refactor : 실패시 alert 창의 스타일 변경

각각 파일의 공백제거
SubViewController 메서드 순서 변경

Commit:
rename : ViewController, SubViewController 파일 이름 변경

ViewController -> MainViewController
SubViewController -> ManagingStockViewController

cherrishRed committed:
add : 앱 아이콘 파일 추가

Commit:
refactor : moveStockChangeView() 함수 기능 분리 및 네이밍 변경

moveStockChangeView() -> moveManagingStockView()
setUpDelegate 함수 분리

cherrishRed committed:
refactor : clickStockChangeButton 네이밍 변경

clickStockChangeButton -> touchUpMoveButton
ManagingStockViewController의 접근제어 설정

Commit:
refactor : stockChangeView 파라미터 네이밍 변경

stockChangeView -> ManagingStockView

cherrishRed committed:
refactor : CustomStringConvertible 출력 부분 방식 변경

Siwon-L committed:
docs : 프로젝트 소개 글 작성

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/MainViewController.swift:
```diff
@@ -53,23 +56,14 @@ class ViewController: UIViewController {
-        } else if fruitName == .pineapple {
-            pineappleLabel.text = String(fruitStock)
-        }
+    private func updateStockLable() {
```
오타가 있네요 ㅎㅎ

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/MainViewController.swift:
```diff
@@ -53,23 +56,14 @@ class ViewController: UIViewController {
-        } else if fruitName == .pineapple {
-            pineappleLabel.text = String(fruitStock)
-        }
+    private func updateStockLable() {
+        strawberryLabel.text = String(juiceMaker.fruitStore.getStock(of:.strawberry))
+        bananaLabel.text = String(juiceMaker.fruitStore.getStock(of:.banana))
+        magoLabel.text = String(juiceMaker.fruitStore.getStock(of:.mango))
+        kiwiLabel.text = String(juiceMaker.fruitStore.getStock(of:.kiwi))
+        pineappleLabel.text = String(juiceMaker.fruitStore.getStock(of:.pineapple))
     }
```
`juiceMaker.fruitStore.getStock`이 너무 중복됩니다. 
`juiceMaker.getStock(of:)` 으로 불러오거나 
`juiceMaker.fruitStore` 를 지역 변수로 지정하면 중복을 줄일 수 있을 것 같습니다.

cherrishRed replied:
네 지역 변수로 처리하였습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/MainViewController.swift:
```diff
@@ -79,33 +73,37 @@ class ViewController: UIViewController {
+        moveManagingStockView()
+    }
+    
+    private func setUpDelegate(_ ManagingStockView: ManagingStockViewController) {
+        ManagingStockView.stock = juiceMaker.fruitStore.stock
+        ManagingStockView.delegate = self
+    }
```
1. `setUpDelegate` 함수의 매개변수가 `ManagingStockViewController` 가 아니면 사용할 곳이 없을 것 같습니다.
어떻게 하면 더 다양한 뷰컨을 받을 수 있을까요?

2. 그리고 따로 함수로 빼신 이유도 궁금합니다!!

3. Main -> Stock은 직접 재고를 전달해 주었고
Stock -> Main 은 델리게이트로 전달해 주셨습니다. 각각 장단점이 궁금합니다!

cherrishRed replied:
1. 이번 리펙토링 과정에서, 프로토콜을 타입으로 지정해 다양한 뷰 컨트롤러를 받을 수 있도록 두개의 메서드로 분리 하였습니다. 
2. 이 메서드가 두군데서 활용되기 때문에 메서드를 따로 빼 놓았습니다. 
3. 우선 각기 다르게 값을 전달한 이유가 있었습니다. main 뷰에서 manage 뷰로 이동시에는 딜리게이트를 사용할 이유가 없어 값을 직접 전달해 주었고, manage 뷰에서 main 뷰로 이동 할 때에는 뷰의 생명주기 때문에 (모달로 창을 띄웠기 때문에 다시 main 으로 돌아올 때에 viewWillAppear 가 실행되지 않는다는 점에서 manage 뷰에서 이 사항을 지시해야 했습니다.)
**직접 값를 전달 할때의 장단점**
장점 - 값을 직접적을 전달하기 때문에 코드가 간겷합니다. 
단점 - 객체간의 결함도가 높아 집니다. 
**딜리게이트를 사용할 때 장단점** 
장점 - 대리자 객체의 생명주기가 아니라, 위임자 객체의 생명주기를 중심으로 일을 시킬 수 있다. 
단점 - 프로토콜을 사용해야 하고, nil 값이 들어갈 경우 오류가 발생할 수 있습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/ManagingStockViewController.swift:
```diff
@@ -0,0 +1,93 @@
+    func update(for stock: [Fruit: Int])
+}
+
+class ManagingStockViewController: UIViewController {
```
재고에 대해서 다양한 화면이 있는 경우가 아니면 그냥 StockVC로 해도 무방해 보입니다!! 어떻게 생각하시나요?

Siwon-L replied:
찾아보니, VC라는 표현을 많이 사용하는것 같습니다.

stevenkim18 replied:
아! 제가 말씀 드린 의도는 VC가 아니라 Managing를 뺴는 경우를 여쭤본거에요 ㅎㅎ

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/ManagingStockViewController.swift:
```diff
@@ -0,0 +1,93 @@
+    @IBOutlet weak var mangoStepper: UIStepper!
+    
+    @IBAction func matchLabel(with stepper: UIStepper) {
+        switch stepper {
+        case strawberryStepper:
+            strawberryLabel.text = String(format: "%.0f", strawberryStepper.value)
+            stock?[.strawberry] = Int(strawberryStepper.value)
+        case bananaStepper:
+            bananaLabel.text = String(format: "%.0f", bananaStepper.value)
+            stock?[.banana] = Int(bananaStepper.value)
+        case pineappleStepper:
+            pineappleLabel.text = String(format: "%.0f", pineappleStepper.value)
+            stock?[.pineapple] = Int(pineappleStepper.value)
+        case kiwiStepper:
+            kiwiLabel.text = String(format: "%.0f", kiwiStepper.value)
+            stock?[.kiwi] = Int(kiwiStepper.value)
+        case mangoStepper:
+            mangoLabel.text = String(format: "%.0f", mangoStepper.value)
+            stock?[.mango] = Int(mangoStepper.value)
+        default:
+            return
+        }
```
- `String(), Int()`가 반복됩니다! 어떻게 반복을 줄일 수 있을까요??
- 실수를 String으로 바꾸는 방법이 format 매개변수를 사용하는 방법 밖에는 없나요?

cherrishRed replied:
딕셔너리에 stepper 와 label 값을 담아 switch 문을 삭제하는 방향으로 코드를 수정하였습니다.

stevenkim18 reviewed (commented):
사파리 레드 마지막 스탭까지 정말 고생많으셨습니다!!

3주 동안 배우신 것들 쭉 한번 복습해보시고 꼭 정리해보시기를 바래요!!

기초가 되는 중요한 것들입니다!!

### ****컨트롤러 부분이 너무 방대해 지는 것에 대해서****

- 자연스러운 현상입니다!!
- 뷰컨에서 많은 일을 하고 있기 때문에 나중에 유지보수하기도 어렵고 테스트 하기도 어려워집니다.
- 그래서 많은 개발자들이 분리하려고 뷰컨을 가볍게 하려고 수많은 노력을 했는데 이는 차차 배우실거에요 ㅎㅎ 이런 고민 자체를 가지고 계신다는 것도 훌룡합니다.
- **지금 단계에서는 함수를 기능 별로 extension으로 분리해주면 좋을 것 같아요**

### ****Stepper의 사이즈가 줄어들지 않는 것에 대해서****

- 기본적인 설정 값으로는 줄어들지 않는 것 같아요.
- UIView를 상속 받고 있어서 사이즈를 조절하는 값을 설정하면 크기가 바뀔 것 같습니다. 만약 되지 않는다면 조금 더 low 단에서 CA가 들어간 함수로 사이즈 조정이 가능합니다!

👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍👍

앱에 아이콘을 넣으신 것은 정말 잘하셨습니다!!!

파일 이름도 그렇고 아이콘도 그렇고 이번 프로젝트에 애정이 가득해 보여서 너무 좋아요!! ㅎㅎ

공부를 해보시는김에 시간이 되시면 왜 이미지 asset에 1x 2x 3x가 필요한지도 한번 찾아보세요 ㅎㅎ

cherrishRed committed:
del : 사용하지 않는 에러 타입 삭제

updateStockLable 오타 Label 로 변경

Commit:
improve : setUpDelegat 함수의 파라미터 결합도를 낮추고, 확장성 증가

Delegator 프로토콜을 채택한 UIViewController를 모두 받을 수 있도록 수정
View 관련 네이밍 축약

cherrishRed committed:
refactor : stepper 와 label 을 묶는 uiGroup 생성

Commit:
refactor : setUpViewValue함수 로직 변경

cherrishRed committed:
refactor : matchLabel 함수 로직 간결하게 변경

tag를 사용하여 stepper 구분

Commit:
style : Updateable프로토콜 MainViewController파일 내부로 이동

공백 제거

cherrishRed committed:
refactor : getStock 반환 타입 변경

updateStockLabel 메서드 축약

Commit:
feat : 재고 수정시 얼럿 알림 추가

cherrishRed committed:
refactor : checkAll 함수 생성 , make 접근제어 변경

Commit:
feat : 품절시 버튼 색 변경해주는 noticeOutOfStock() 함수 추가

checkAll() 로직 변경
matchJuice() 함수 삭제
buttonGroup :[UIButton: Juice] 추가

Siwon-L commented:
@stevenkim18

안녕하세요 steven 😁 🤚🏻
벌써 프로젝트가 끝이 나고 있네요🥲

STEP2 리펙토링 완료 했습니다.
저희가 개인적으로 프로젝트에 넣고 싶은 기능도 추가 했어요!😁

# UML
<img src="https://user-images.githubusercontent.com/91936941/157674726-ebd2f28a-e6bd-4d4f-b2b9-ece24e5a7f76.png" width="1000">

# 변경 내용 및 질문
- 변경한 부분
    - 오타 및 사용하지 않는 코드 제거
    - 중복되는 코드 축약
    - setUpDelegat 함수의 파라미터 결합도를 낮추고, 확장성을 높이기
    - 뷰컨트롤러 네이밍 수정
    - String Int 타입 변환 반복 줄이기
    - 재고품절시 버튼 백그라운드 컬러 변경 기능 추가
    - 재고 수정시 확인 얼럿창 기능 추가
- 조언을 얻고 싶은 부분
    - UIview tag 사용에 대해 
    - 함수를 기능 별로 extension으로 분리하는 것에 대해 
    - UML에 관하여 

## UIview tag 사용에 대해 
matchLabel() 함수의 로직을 간결하게 하기 위해서 rawValue와 tag를 사용하게 되었습니다. 하지만, 사용하며 tag의 값은 정수이고, 스토리보드에서 직접적으로 설정하여 코드의 가독성을 해치는 것과 하드코딩이 될 수 있지 않을까라는 생각이 들었습니다. 때문에 이 부분에 대해 스티븐의 생각을 듣고 싶습니다.

## 함수를 기능 별로 extension으로 분리하는 것에 대해 
> 지금 단계에서는 함수를 기능 별로 extension으로 분리해주면 좋을 것 같아요

라고 조언을 남겨 주셨는데 저희가 어떤 방식을 steven 이 설명하시는지 잘 이해하지 못한 것 같습니다.😂

## UML에 관하여
UML에 대해 공부하였습니다.
하지만, UML은 너무 자바 위주로 만들어져 있어 swift 프로젝트를 UML로 만드는 부분에서 많은 어려움을 느꼈습니다. 상단에 저희가 활동 학습 시간에 만든(프로젝트 수정 전) UML를 첨부하였고, 직접 작성하며,  
- 열거형 타입이 전역으로 넓게 사용되는 경우 모든 객체와 연결을 해줘야 하는 부분
- 딜리게이트 패턴일때 표현 부분
- extension의 표현 부분
위와 같은 상황에서 어떤 UML적 표현을 해야 할지 잘 모르겠습니다.

### asset 이미지에 1x 2x 3x가 왜 필요한지에 대해서 
알아보라고 하셔서 공부를 해보았습니다. 
애플은 기기에 따라 레티나 디스플레이를 사용하는 기기도 있고 아닌 화면도 있다고 알고 있습니다. 
레티나 디스플레이가 아닐 경우에는 1point == 1pixel 
레티나 디스플레이의 경우 1point == 4pixel 
HD 레티나 디스플레이의 경우 1point == 9pixel 
입니다. 
그래서 해상도 별로 1x 2x 3x 의 다양한 이미지가 필요합니다. 
물론 3x 이미지를 줄여 사용하거나(다운 스케일) 1x 이미지를 늘려 사용하는 방법(업 스케일)도 있으나, 다운 스케일이나 업스케일을 사용하면 이미지가 깨질 수 있어 다양한 크기의 이미지를 준비하는 것이 좋습니다.

Siwon-L committed:
docs : REDME 수정

Siwon-L committed:
docs : README 수정

Siwon-L committed:
docs : README 수정

stevenkim18 commented:
### UIview tag 사용에 대해
- 코드의 가독성과 간결성은 항상 충돌되는 부분입니다. 각각 일장일단이 있기 때문에 정답은 없습니다. 떄에 따라 가장 최선을 생각하시면 됩니다.
- 다양하게 시도해보시라고 드린 코멘트였습니다 ㅎㅎ 합당한 이유가 있으시면 꼭 줄이실 필요는 없었습니다.
- 저도 uiview tag는 잘 사용하지만 않지만 rawValue에 값을 정의하고 Fruit에 연산프로퍼티나 메소드를 잘 활용하면 조금은 간결함안에 가독성이 올라갈 수 있을 것 같습니다.

### 함수를 기능 별로 extension으로 분리하는 것에 대해
뷰컨에 너무 많은 메소드가 있으시다고 하셔서 기능 별로 분리를 요청드린거였습니다.
```swift
class MainViewController: UIViewController {
    override func viewDidLoad()
    func update(for stock: [Fruit: Int])
    ...
}

// MARK: 뷰 관련, 재고 관련
extension MainViewController {
    private func setUpDelegate(_ viewNavigation: UINavigationController)
    private func moveManagingStockView() -> UINavigationController?
    ....
}
```
- extension은 기능을 확장할 때 주로 사용하지만 함수를 나눌때도 사용합니다!

### UML에 관하여
- 전역변수 일때 -> 그냥 전역변수 영역을 만들어서 거기 안에 둘 것 같습니다.
- 딜리게이트 패턴 -> 자바에서 인터페이스가 swift에서는 프로토콜 입니다. 프로토콜을 채택하고 관련 함수를 사용했다고 표시하면 될 것 같습니다.
- extension -> 저도 확장한다는 의미에서 화살표를 사용할 것 같습니다.
아니면

![](https://i.stack.imgur.com/u9nkx.png)
이런 느낌으로 객체 안에도 넣을 수 있을 것 같아요!!

--- 
레드 사파리 3주 동안 너무 고생많으셨습니다.
저의 담당은 끝이 났지만 언제든지 궁금한점 있으시면 디엠 보내주세요!!

cherrishRed committed:
refactor : showFailureAlert 함수 네이밍 변경

showFailureAlert -> askForChangeAlert
noticeOutOfStock 메서드의 접근제어자 붙이기

Commit:
refactor : 얼럿기능 extension으로 분리

ManagingStockViewController -> StockViewController 네이밍 변경
managing네임 삭제 및 VN -> NC로 변경

cherrishRed committed:
docs : ReadMe의 UML 수정

Commit:
rename : ManagingStockViewController.swif -> StockViewController.swift 변경

cherrishRed committed:
docs : ReadMe 목차 수정
